### PR TITLE
feat: decouple Pi0.5 policy into standalone module

### PR DESCRIPTION
**Pull request by eugene123tw**
# Pull Request

## Description

Decouple Pi0.5 policy into its own standalone module under `library/src/physicalai/policies/pi05/`.

Previously, Pi0.5 was tightly coupled with the Pi0 implementation (`policies/pi0`). This PR extracts it into a self-contained package with dedicated modules for configuration, model architecture, preprocessing, Gemma integration, and pretrained weight loading. The decoupled design makes Pi0.5 easier to maintain, test, and extend independently of Pi0.

## Type of Change

- [x] ✨ `feat` - New feature
- [x] ♻️ `refactor` - Code refactoring
- [x] 🧪 `test` - Tests

## Breaking Changes

- `Pi05` and `Pi05Config` are now imported from `physicalai.policies.pi05` instead of `physicalai.policies.pi0`. The top-level `physicalai.policies` re-exports remain unchanged, so code using `from physicalai.policies import Pi05` continues to work.
- New export `Pi05Model` is added to the public API.

---

### New module: `library/src/physicalai/policies/pi05/`

| File | Lines | Purpose |
|------|-------|---------|
| `__init__.py` | 10 | Package exports (`Pi05`, `Pi05Config`, `Pi05Model`) |
| `config.py` | 122 | `Pi05Config` dataclass with validation — flow matching hyperparameters, optimizer/scheduler settings, compilation options |
| `model.py` | 916 | `Pi05Model` — dual-backbone architecture (PaliGemma VLM + Gemma action expert), flow matching forward/inference, sinusoidal position encoding, KV cache support |
| `pi_gemma.py` | 341 | Extended Gemma/PaliGemma with adaptive RMSNorm (AdaRMS) and gated residuals for the action expert |
| `policy.py` | 475 | `Pi05` Lightning wrapper — dual-path init (lazy via `trainer.fit()`, eager via `load_from_checkpoint()`), training step, inference, pretrained weight loading from HuggingFace |
| `preprocessor.py` | 313 | `Pi05Preprocessor` / `Pi05Postprocessor` — state normalization and tokenization, image resize+pad, action normalization/padding, language tokenization |
| `pretrained_utils.py` | 240 | Utilities for loading lerobot-format pretrained weights — normalization stat conversion (QUANTILES/MIN_MAX → MEAN_STD), state dict key remapping, feature shape resolution |

### Modified files

- **`policies/__init__.py`** — Import `Pi05`, `Pi05Config`, `Pi05Model` from new `pi05` module instead of `pi0`
- **`gyms/libero.py`** — Pass `task_description` in observation dict so Pi0.5 can build language prompts
- **`tests/unit/gyms/test_libero.py`** — Two new tests verifying task description is present in reset/step observations
- **`tests/integration/test_first_party_e2e.py`** — Add `"pi05"` to `FIRST_PARTY_VLA_POLICIES` and configure its E2E test with `freeze_vision_encoder=True, train_expert_only=True`
- **`tests/unit/policies/test_pi05.py`** — 851 lines of unit tests covering config validation, model construction, preprocessor, pretrained utils, and policy API

### Summary

- **+3,299 / −2** lines across 12 files
- 11 commits on `feat/pi-0.5-decouple`

### API usage

```python
from physicalai.policies.pi05 import Pi05, Pi05Config

# Default config
config = Pi05Config()

# Custom config
config = Pi05Config(
    paligemma_variant="gemma_2b",
    action_expert_variant="gemma_300m",
    chunk_size=50,
    num_inference_steps=10,
    freeze_vision_encoder=True,
    train_expert_only=True,
)

# Load pretrained from HuggingFace
policy = Pi05(pretrained_name_or_path="lerobot/pi0.5_base")

# Load pretrained from HuggingFace
policy = Pi05(pretrained_name_or_path="lerobot/pi05_libero_finetuned")

policy.to("cuda")

# Evaluate on LIBERO (equivalent to your lerobot eval)
benchmark = LiberoBenchmark(
    task_suite="libero_10",   # --env.task=libero_10
    task_ids=[0],             # --env.task_ids=[0]
    num_episodes=1,           # --eval.n_episodes=1
    max_steps=500,
    video_dir="videos/pi05_libero",  # save videos here
    record_mode="all",               # "all", "successes", "failures", or "none"
)
results = benchmark.evaluate(policy)
print(results.summary())
```

### CLI usage

```bash
physicalai fit --config configs/physicalai/pi05.yaml
```

### CODE REVIEW

File: library/src/physicalai/policies/pi05/config.py [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+    time_sampling_scale: float = 0.999
+    time_sampling_offset: float = 0.001
+    min_period: float = 4e-3
+    max_period: float = 4.0
+
+    image_resolution: tuple[int, int] = (DEFAULT_IMAGE_SIZE, DEFAULT_IMAGE_SIZE)
+
+    empty_cameras: int = 0
+
+    tokenizer_max_length: int = 200
```
[the remark this answers is not in the document]
Comment (sovrasov) [reply library/src/physicalai/policies/pi05/config.py]: I see, ok then, so it's valid for pi0 config

File: library/src/physicalai/policies/pi05/preprocessor.py [hunk 2]
```diff
@@ -0,0 +1,405 @@
...
+
+logger = logging.getLogger(__name__)
+
+NORM_MAP = {
+    FeatureType.STATE: NormalizationType.MEAN_STD,
+    FeatureType.ACTION: NormalizationType.MEAN_STD,
+}
+
+
+def pad_vector(vector: Tensor, new_dim: int) -> Tensor:
```
Comment (samet-akcay) [inline library/src/physicalai/policies/pi05/preprocessor.py]:
```suggestion
def pad_vector(vector: torch.Tensor, new_dim: int) -> torch.Tensor:
```

**Review by samet-akcay — CHANGES_REQUESTED** [review]
great pr, thanks for the effort! only super minor comments.

File: library/src/physicalai/policies/pi05/policy.py [hunk 3]
```diff
@@ -0,0 +1,489 @@
+# Copyright (C) 2026 Intel Corporation
+# SPDX-License-Identifier: Apache-2.0
+
+# Copyright 2025 Physical Intelligence and The HuggingFace Inc. team.
```
[the remark this answers is not in the document]
Comment (eugene123tw) [reply library/src/physicalai/policies/pi05/policy.py]: Updated
